### ISSUE TITLE
[Concept] The Workshop

### ISSUE THREAD

**Issue by Aeltumn** [Aeltumn/Chisels-and-Bits-2#14]
I saw this name in a reddit thread first, so credits to u/Matalya1: https://www.reddit.com/r/ModdedMinecraft/comments/ct0rej/idea_for_chisel_bits_workshop/

Anyways, here something that I might add to C&B2 in the future and I'd like to hear feedback from players to make it the best it can be.

The workshop would be a combination between a website and in-game blueprints where players could freely upload their in-game multiblock creations to the internet and even preview creations of others in-game! I think the Sims 4 gallery is the most similar to how I'd imagine this workshop, integrated directly allowing you to easily share your work.

The amount of downloads a build gets could even be a metric of populartiy to showcase cool creations.

I'm interested in adding more sharing functionality to C&B2 instead of many features other mods (like LittleTiles) already offer.

So, I'm interested in community feedback, what functionality would this system really need? How big should the maximum size of an uploadable creation be? How should selecting your build work: with a tool that looks like drap region/tape measure or a machine with GUI? Should a button to view creations of others work by having a hotkey (like advancements) that opens a full-screen web page, or should you download files from the website, put them in a folder and select which one to build in a GUI similar to the stonecutter/loom? Maybe a new button should be added to the escape menu which opens the web page on which you can click the download button to download files which are automatically scanned by the mod to appear ingame? Maybe there could be a new item where all recently downloaded designs are shown in a radial menu where you can select which one you want to place? Should you be able to place only parts of a downloaded creation, maybe creators should be able to mark certain parts as optional parts of the creation to upload/place them in segments?

Many questions indeed, but there's still some time to figure out the optimal way for this workshop system to work, it's only going to get added after everything else is done, so that's still a good few months out.

Personally, I prefer options to keep what is done ingame as vanilla-esque as possible, I want to make things in way where it'd be believable for this to be added in a future minecraft update and maybe more importantly: it needs to be intuitive and "noob-proof".

**Comment by MPNanKing** [comment]
Will version 1.12 be included in this plan? If the mod is not modified, blocks can be in the form of a clipboard file(chiselsandbits_clipboard.cfg) in old version.

**Comment by Aeltumn** [comment]
> 
> 
> Will version 1.12 be included in this plan? If the mod is not modified, blocks can be in the form of a clipboard file(chiselsandbits_clipboard.cfg) in old version.

No, the workshop won't be for 1.15, it might even not get added until 1.16 comes around. But I am planning on adding some compatibility code to convert old 1.12 worlds to 1.14+.

**Comment by Tompouce-onGit** [comment]
Having discovered the existence of this mod and FCB just recently, and subsequently the retirment of the original author, I am very excited for your rework of the mod to release ! 😃
I have never played the original mod (since I only discovered it some days ago), but since you emphasized that you'd like to make it as intuitive for new players as possible, I think I can still give some suggestions of what players using the mod for the first time could expect!

- First of all, I agree that controling the mod with in-game interactions as much as possible would be very nice. That is why I'm not certain making the web page you are envisioning accessible through the settings screen would be fitting ; usually, players open the pause menu when they are looking to do something that's not related to gameplay, like changing audio and video settings, or just stepping away from the game for a moment. I don't think this functionnality would fit there.

- I believe having an item like a book (named something like 'creations/models catalogue' for instance), or a workstation-like block, from which you can view the creations you have downloaded and access this sharing website, via a link or something, would feel right. In my opinion, it's a nice way to access this functionnality in-game and intuitively. Maybe one could even browse user-made models from this item/block directly, i.e. it could fetch download links and thumbnails of recent or popular creation from the website and display them in a GUI. Although I don't know how practical it would be to implement, or how much work it would require 😕. I have never made mods for minecraft myself so...

- As for potential hotkeys, I reckon trying to keep the number of additionnal key bindings to a minimum. I have two reasons for this : the first is that I find that modpacks are plagued with the issue of many mods lots of hotkeys, causing many conflicts that are a pain to resolve since there are only so many keys on a keyboard. If this mod is to be added to modpacks in the future (which I can totaly see happening), then reducing the ammount of hotkeys would do a lot to reduce this issue. The second reason is that, in my opinion, new players that are just getting into modding their minecraft client might not realise immediately that some functionnality is accessible via another key. Items and blocks are directly visible and can be directly interacted with, whereas you have to know that there is a hotkey in order to use it (Although I guess such hotkeys could be explained in-game in some way). In any case, if the functionnality of viewing you own and downloaded creations is integrated into an item or block, I don't see the necessity of adding hotkeys.

- finally, having a way to label creations that you have made or downloaded as favourite, or even organising them into different categories that you could name would be a nice way to make it easy and intuitive to quickly get access to a creation that you want to use.

As for the maximum size of an uploadable creation, I really need to play around with the mod to get an idea.

I hope my suggestions were useful to you in some way, and I can't wait to see what you do in the end!
I also hope that you and your entourage are safe durring the lockdown, and I wish you all the best for your debut in the modding scene 😄

**Comment by Aeltumn** [comment]
Thanks for the feedback, it'll probably be a while before I start working on the Workshop but when I'll do I keep your feedback in mind.

About hotkeys, C&B has most of its hotkeys unbound by default, they are all optional. Because I agree, most modpacks have insane hotkey overlap problems. Some mods even have keybind conflicts with vanilla, people put too little effort into keybindings.

**Comment by Tompouce-onGit** [comment]
Glad to hear that!

**Comment by LordMZTE** [comment]
It sounds like an awesome concept! but it brings a lot of questions...
about the selecting of an area dragging a box does seem like a good idea to me.
the uploading could be a block or maybe an item. the block could be placed near the selection and the item could simply be clicked near the selection. they would have some input fields for a name and a description, maybe a 3D render of the selection and an upload button. Once the upload button has been clicked and information has been entered a link could appear in chat going to the creation. to download maybe you would have a similar item that has an input field for an ID which can be found on the website for each creation and a Load button, then it would behave as a blueprint would.
another issue is editing the title and the description after uploading and taking credit for creations. how could you prove an uploaded creation is yours? of course, this can be solved with accounts but that sounds unnecessary to me if there was a smarter solution. maybe, the uploaded blueprints could have the in game name of the player who uploaded them shown on their page. but i have no idea how that could prove that a player is the auther of a creation for editing title and description since it seems rather trivial to change the name that will appear by changing the mod's code slightly or (maybe) by some really aggressive nickname mods (not sure how realistic that is honestly).
I am not a fan of the idea of having a full on web browser in game as you suggested. much rather I think the mod should have a custom menu that talks to the website via an API. but I'm not sure if it is even necessary to have a menu in game instead of just a button that links to a public list on the website.
another question that comes to mind is how the creations will be serialized. 
the first idea I had about this was using JSON. the file could simply contain an array of objects which each have a position and an ID for which block they are. this would make it simple to work with and it would be easy to check which mods are required however this has the disadvantage of being rather uncompressed for big shapes of identical bits. for that, a solution might be to have each entry in the JSON file have 2 sets of coordinates that correspond to a starting and an end position. the cube that they would make up could then be filled with the right bit.
another advantage of json files is that it would be easy to include additional information like name and description.
alternatively, maybe things like obj file could be used? I'm not sure how those work and if it would be possible to save the materials the bits have or additional data in them but if so, they would be even better.
